lib/rucio/daemons/common.py

```diff
@@ -18,27 +18,143 @@
+        """
+        self.once: Final[bool] = once
+        self.total_workers: Final[int] = total_workers
+        self.sleep_time: Final[int] = sleep_time
+        self.partition_wait_time: Final[int] = partition_wait_time
+        self.daemon_name = daemon_name
+        self.graceful_stop = threading.Event()
+        setup_logging(process_name=daemon_name)
+
+    @staticmethod
+    def _pre_run_checks() -> None:
+        """
+        Checks to run before daemon execution
+        """
+        if rucio.db.sqla.util.is_old_db():
+            raise DatabaseException("Database was not updated, daemon won't start")
+
+    @abstractmethod
+    def _run_once(
+        self, heartbeat_handler: "HeartbeatHandler", **_kwargs
+    ) -> tuple[bool, Any]:
+        """
+        Daemon-specific logic (to be defined in child classes) for a single iteration
+
+        :param heartbeat_handler: Handler to set and manage the heartbeat for this execution.
+
+        :returns: Tuple of (must_sleep, ret_value).
+                  must_sleep: set to True to signal to the calling context that it must sleep before next execution, False otherwise
+                  ret_value: Daemon-specific return value
+        """
+        pass
+
+    def _call_daemon(self, activities: Optional[list[str]] = None) -> None:
+        """
+        Run the daemon loop and call the function _run_once at each iteration
+        """
+
+        while not self.graceful_stop.is_set():
+            run_once_fnc = functools.partial(self._run_once)
+
+            daemon = db_workqueue(
+                once=self.once,
+                graceful_stop=self.graceful_stop,
+                executable=self.daemon_name,
+                partition_wait_time=self.partition_wait_time,
+                sleep_time=self.sleep_time,
+                activities=activities,
+            )(run_once_fnc)
+
+            for _ in daemon():
+                pass
+
+    def run(self) -> None:
+        """
+        Run the daemon.
+        """
+        self._pre_run_checks()
+
+        logging.info("%s: starting threads", self.daemon_name)
+        thread_list = [
+            threading.Thread(target=self._call_daemon)
+            for _ in range(self.total_workers)
+        ]
+
+        for t in thread_list:
+            t.start()
+
+        for t in thread_list:
+            t.join()
+
+        if not self.once:
+            logging.info("%s: waiting for interrupts", self.daemon_name)
+
+    def stop(
+        self, signum: Optional[int] = None, frame: Optional[FrameType] = None
+    ) -> None:
+        """
+        Graceful exit the daemon. Used as handler for SIGTERM.
+        The unused parameters are needed for the method to be accepted as handler for signal.signal().
+
+        :param signum: signal number
+        :param frame: stack frame
+        """
+        self.graceful_stop.set()
+        logging.info("%s: terminating", self.daemon_name)
+        if signum == signal.SIGINT or signum == signal.SIGTERM:
```

Comment:
should be sigint only, not sigterm as well

Reply:
worth noting: the daemons in their current version use `SIGTERM`, not `SIGINT`, e.g. https://github.com/rucio/rucio/blob/c8316a595af9ef6336a131ead947414f32e7a8ca/bin/rucio-auditor#L86

Reply:
The Auditor is not a good example.

`SIGTERM` is what the OS/supervisor/scheduler is using to terminated a process under its control. It is supposed to send `SIGKILL` after a timeout.

`SIGINT` is what Ctrl+C sends. It only makes sense when one starts a daemon process from their terminal, which is unlikely, especially in a production environment.

The daemons that are (supposedly) handling `SIGNIT` are doing so via the `KeyboardInterrupt` exception.

Reply:
I linked to Auditor as a random example, but it was part of this search https://github.com/search?q=repo%3Arucio%2Frucio%20sigterm&type=code where it seems like most daemons use `SIGTERM`. I agree that for the purpose of gracefully stopping manually we should use `SIGINT`, I just wanted to bring this up for reference